internal/state/change_processor.go

```diff
@@ -75,40 +82,135 @@ func (c *ChangeProcessorImpl) CaptureUpsertChange(obj client.Object) {
 
 	switch o := obj.(type) {
 	case *v1beta1.GatewayClass:
-		if o.Name != c.cfg.GatewayClassName {
-			panic(fmt.Errorf("gatewayclass resource must be %s, got %s", c.cfg.GatewayClassName, o.Name))
-		}
-		// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-		if c.store.gc != nil && c.store.gc.Generation == o.Generation {
-			resourceChanged = false
-		}
-		c.store.gc = o
+		resourceChanged = c.captureGatewayClassChange(o)
 	case *v1beta1.Gateway:
-		// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-		prev, exist := c.store.gateways[getNamespacedName(obj)]
-		if exist && o.Generation == prev.Generation {
-			resourceChanged = false
-		}
-		c.store.gateways[getNamespacedName(obj)] = o
+		resourceChanged = c.captureGatewayChange(o)
 	case *v1beta1.HTTPRoute:
-		// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-		prev, exist := c.store.httpRoutes[getNamespacedName(obj)]
-		if exist && o.Generation == prev.Generation {
-			resourceChanged = false
-		}
-		c.store.httpRoutes[getNamespacedName(obj)] = o
+		resourceChanged = c.captureHTTPRouteChange(o)
+	case *v1.Service:
+		resourceChanged = c.captureServiceChange(o)
+	case *discoveryV1.EndpointSlice:
+		resourceChanged = c.captureEndpointSliceChange(o)
 	default:
 		panic(fmt.Errorf("ChangeProcessor doesn't support %T", obj))
 	}
 
 	c.storeChanged = c.storeChanged || resourceChanged
 }
 
+func (c *ChangeProcessorImpl) captureGatewayClassChange(gc *v1beta1.GatewayClass) bool {
```

Comment:
I moved the methods to the store object.

I also moved the `storeChanged` bool from the change processor to the store, but I didn't add a `Changed()` method. Since the store and the change processor are in the same package I don't think it's necessary.